src/core/DY.Core.Trace.Arithmetic.Derived.fst

```diff
@@ -49,6 +49,182 @@ let trace_subtract_concat_slices tr1 tr2 tr3 =
   trace_subtract_concat tr2 tr3
 
 
+/// Properties relating trace concatenation and subtraction to trace search
+
+#push-options "--z3rlimit 50"
+val trace_search_first_concat:
+  #label_t:Type ->
+  tr1:trace_ label_t -> tr2:trace_ label_t ->
+  p:(trace_entry_ label_t -> bool) ->
+  Lemma
+  (ensures
+    (trace_search_first (tr1 <++> tr2) p) == (
+      match trace_search_first tr1 p with
+      | None -> (
+        match trace_search_first tr2 p with
+        | None -> None
+        | Some ts -> Some (ts + trace_length tr1 <: timestamp)
+      )
+      | Some ts -> Some ts
+    )
+  )
+let trace_search_first_concat #label_t tr1 tr2 p =
+  match trace_search_first tr1 p with
+  | None -> (
+    match trace_search_first tr2 p with
+    | None -> introduce forall (ts:timestamp). ts `on_trace` (tr1 <++> tr2) ==> ~(p(get_entry_at (tr1 <++> tr2) ts))
+             with introduce _ ==> _
+             with _. trace_concat_get_entry tr1 tr2 ts
+    | Some ts -> (
+      trace_concat_get_entry tr1 tr2 (ts + trace_length tr1);
+      introduce forall (ts':timestamp). ts' `on_trace` (tr1 <++> tr2) /\ p(get_entry_at (tr1 <++> tr2) ts') ==> ts' >= ts + trace_length tr1
+      with introduce _ ==> _
+      with _. trace_concat_get_entry tr1 tr2 ts'
+    )
+  )
+  | Some ts -> assert(tr1 <$ (tr1 <++> tr2)) // Triggers trace_search_first_later
```

Comment:
Interesting, I wonder whether this is a hint toward relaxing the SMT pattern of `trace_concat_grows` to `SMTPat (tr1 <++> tr2)` insteaf of `SMTPat (tr1 <$ (tr1 <++> tr2))`

Reply:
Discussed below, this is not too bad because it's in the library, we might revise the SMT pattern if such asserts happen too much in user code

Reply:
(hence resolving the conversation)